src/libsystemd/sd-journal/journal-file.c

```diff
@@ -552,18 +552,103 @@ static int journal_file_verify_header(JournalFile *f) {
         if (UINT64_MAX - header_size < arena_size || header_size + arena_size > (uint64_t) f->last_stat.st_size)
                 return -ENODATA;
 
-        if (le64toh(f->header->tail_object_offset) > header_size + arena_size)
+#define check_offset_basic(offset)                              \
+        {                                                       \
+                uint64_t _p = (offset);                         \
+                                                                \
+                if (!VALID64(_p))                               \
+                        return -ENODATA;                        \
+                if (_p != 0 && _p < header_size)                \
+                        return -ENODATA;                        \
+        }
+
+#define check_offset(offset)                    \
+        {                                       \
+                uint64_t _q = (offset);         \
+                                                \
+                check_offset_basic(_q);         \
+                if (_q > tail_object_offset)    \
+                        return -ENODATA;        \
+        }
+
+#define check_hash(name)                                                \
+        {                                                               \
+                uint64_t _h, _sz;                                       \
+                                                                        \
+                _h = le64toh(f->header->name##_hash_table_offset);      \
+                _sz = le64toh(f->header->name##_hash_table_size);       \
+                if ((_h == 0) != (_sz == 0))                            \
+                        return -ENODATA;                                \
+                if (_h != 0) {                                          \
+                        if (_h < offsetof(Object, hash_table.items))    \
+                                return -ENODATA;                        \
+                        _h -= offsetof(Object, hash_table.items);       \
+                        check_offset(_h);                               \
+                        assert(_h <= header_size + arena_size);         \
+                        if (_sz > header_size + arena_size - _h)        \
+                                return -ENODATA;                        \
+                }                                                       \
+        }
```

Comment:
do we really need to make these macros? i.e. the fact that they access local variables directly is particularly ugly.

why not just make them regular static functions, and pass the few variables in as parameters? and then make it return a regular error code?

yes, it would be a bit longer, but a lot less special.